src/librustdoc/html/render/span_map.rs

```diff
@@ -0,0 +1,209 @@
+use crate::clean;
+use crate::html::sources;
+
+use rustc_data_structures::fx::FxHashMap;
+use rustc_hir::def::{DefKind, Res};
+use rustc_hir::def_id::DefId;
+use rustc_hir::intravisit::{self, NestedVisitorMap, Visitor};
+use rustc_hir::{ExprKind, GenericParam, GenericParamKind, HirId, Mod, Node};
+use rustc_middle::ty::TyCtxt;
+use rustc_span::Span;
+
+use std::path::{Path, PathBuf};
+
+/// This enum allows us to store two different kinds of information:
+///
+/// In case the `span` definition comes from the same crate, we can simply get the `span` and use
+/// it as is.
+///
+/// Otherwise, we store the definition `DefId` and will generate a link to the documentation page
+/// instead of the source code directly.
```

Comment:
Why did this need a new type? Rustdoc already has this logic for types because they can be either local or external - how is it done there?

Reply:
The information it holds is simplified. In short, if local, we can use the span to generate the URL to go to the definition, otherwise we use the `DefId` to generate a link to the item's documentation.